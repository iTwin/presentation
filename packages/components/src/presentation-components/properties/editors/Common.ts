/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/

import { PropertyValue } from "@itwin/appui-abstract";

/** @internal */
export interface PropertyEditorAttributes {
  getValue: () => PropertyValue | undefined;
  htmlElement: HTMLElement | null;
}
