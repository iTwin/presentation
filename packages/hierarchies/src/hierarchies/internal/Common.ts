/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/

import naturalCompare from "natural-compare-lite";
import { ConcatenatedValue } from "@itwin/presentation-shared";
import { HierarchyNodeKey } from "../HierarchyNodeKey.js";

/**
 * This is a logging namespace for public log messages that may be interesting to consumers.
 * @internal
 */
export const LOGGING_NAMESPACE = "Presentation.Hierarchies";

/**
 * This is a logging namespace for public performance-related log messages that may be interesting to consumers.
 * @internal
 */
export const LOGGING_NAMESPACE_PERFORMANCE = `${LOGGING_NAMESPACE}.Performance`;

/**
 * This is a logging namespace for internal log messages that are only interesting to package contributors,
 * but not the consumers.
 * @internal
 */
export const LOGGING_NAMESPACE_INTERNAL = "Presentation.HierarchiesInternal";

/**
 * This is a logging namespace for internal performance-related log messages that are only interesting
 * to package contributors, but not the consumers.
 * @internal
 */
export const LOGGING_NAMESPACE_PERFORMANCE_INTERNAL = `${LOGGING_NAMESPACE_INTERNAL}.Performance`;

/** @internal */
export function createOperatorLoggingNamespace(
  operatorName: string,
  baseCategory:
    | typeof LOGGING_NAMESPACE
    | typeof LOGGING_NAMESPACE_PERFORMANCE
    | typeof LOGGING_NAMESPACE_INTERNAL
    | typeof LOGGING_NAMESPACE_PERFORMANCE_INTERNAL,
) {
  return `${baseCategory}.Operators.${operatorName}`;
}

/** @internal */
/* c8 ignore start */
export function createNodeIdentifierForLogging(
  node: { label: string | ConcatenatedValue; key: HierarchyNodeKey; parentKeys?: HierarchyNodeKey[] } | undefined,
) {
  if (!node) {
    return "<root>";
  }
  const { label, key } = node;
  const parentKeys = "parentKeys" in node ? node.parentKeys : "<unknown>";
  return JSON.stringify({ label, key, parentKeys });
}
/* c8 ignore end */

/** @internal */
export function hasChildren<TNode extends { children?: boolean | Array<unknown> }>(node: TNode) {
  return node.children === true || (Array.isArray(node.children) && node.children.length > 0);
}

/** @internal */
export function compareNodesByLabel<TLhsNode extends { label: string }, TRhsNode extends { label: string }>(lhs: TLhsNode, rhs: TRhsNode): number {
  return naturalCompare(lhs.label.toLocaleLowerCase(), rhs.label.toLocaleLowerCase());
}

/**
 * A helper that disposes the given object, if it's disposable. The first option is to dispose
 * using the `Symbol.dispose` method if it exists on the object. If not, fall back to the deprecated
 * `dispose` method if it exists. If not, the object is considered as non-disposable and nothing
 * is done with it.
 *
 * @internal
 */
export function safeDispose(disposable: {} | { [Symbol.dispose]: () => void } | { dispose: () => void }) {
  if ("dispose" in disposable) {
    disposable.dispose();
  } else if (Symbol.dispose in disposable) {
    disposable[Symbol.dispose]();
  }
}
