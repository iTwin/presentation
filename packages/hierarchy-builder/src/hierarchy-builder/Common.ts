/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/

import { ECClass, SchemaContext, SchemaKey } from "@itwin/ecschema-metadata";
import { InProgressTreeNode } from "./TreeNode";

export async function getClass(schemas: SchemaContext, fullClassName: string) {
  const [schemaName, className] = fullClassName.split(/[\.:]/);
  const schema = await schemas.getSchema(new SchemaKey(schemaName));
  if (!schema) {
    throw new Error(`Invalid schema: ${schema}`);
  }
  const nodeClass = await schema.getItem<ECClass>(className);
  if (!nodeClass) {
    throw new Error(`Invalid class: ${nodeClass}`);
  }
  return nodeClass;
}

export function mergeInstanceNodes<TDirectChildren>(
  lhs: InProgressTreeNode,
  rhs: InProgressTreeNode,
  directChildrenMerger: (lhsChildren: TDirectChildren, rhsChildren: TDirectChildren) => TDirectChildren,
): InProgressTreeNode {
  if (lhs.key.type !== "instances" || rhs.key.type !== "instances") {
    throw new Error("Only instance nodes allowed");
  }
  return {
    label: lhs.label,
    key: {
      type: "instances",
      instanceKeys: [...lhs.key.instanceKeys, ...rhs.key.instanceKeys],
    },
    mergeByLabelId: lhs.mergeByLabelId,
    children:
      Array.isArray(lhs.children) && Array.isArray(rhs.children)
        ? [...lhs.children, ...rhs.children]
        : lhs.children === true || rhs.children === true
        ? true
        : lhs.children === false && rhs.children === false
        ? false
        : undefined,
    ...(lhs.hideIfNoChildren && rhs.hideIfNoChildren ? { hideIfNoChildren: lhs.hideIfNoChildren && rhs.hideIfNoChildren } : undefined),
    ...(lhs.hideInHierarchy && rhs.hideInHierarchy ? { hideInHierarchy: lhs.hideInHierarchy && rhs.hideInHierarchy } : undefined),
    ...(lhs.groupByClass || rhs.groupByClass ? { groupByClass: lhs.groupByClass || rhs.groupByClass } : undefined),
    ...(lhs.autoExpand || rhs.autoExpand ? { autoExpand: lhs.autoExpand || rhs.autoExpand } : undefined),
    ...(lhs.extendedData || rhs.extendedData ? { extendedData: { ...lhs.extendedData, ...rhs.extendedData } } : undefined),
    ...(lhs.directChildren || rhs.directChildren ? { directChildren: directChildrenMerger(lhs.directChildren, rhs.directChildren) } : undefined),
  };
}
