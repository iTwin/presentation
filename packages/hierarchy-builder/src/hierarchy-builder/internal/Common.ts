/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/

import { merge, Observable } from "rxjs";
import { assert } from "@itwin/core-bentley";
import { HierarchyNode, HierarchyNodeHandlingParams, HierarchyNodeKey } from "../HierarchyNode";
import { ECClass, ECSchema, IMetadataProvider, parseFullClassName } from "../Metadata";

/** @internal */
export async function getClass(metadata: IMetadataProvider, fullClassName: string): Promise<ECClass> {
  const { schemaName, className } = parseFullClassName(fullClassName);
  let schema: ECSchema | undefined;
  try {
    schema = await metadata.getSchema(schemaName);
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error(e);
  }
  if (!schema) {
    throw new Error(`Invalid schema "${schemaName}"`);
  }

  let nodeClass: ECClass | undefined;
  try {
    nodeClass = await schema.getClass(className);
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error(e);
  }
  if (!nodeClass) {
    throw new Error(`Invalid class "${className}" in schema "${schemaName}"`);
  }

  return nodeClass;
}

function mergeNodeHandlingParams(
  lhs: HierarchyNodeHandlingParams | undefined,
  rhs: HierarchyNodeHandlingParams | undefined,
): HierarchyNodeHandlingParams | undefined {
  if (!lhs && !rhs) {
    return undefined;
  }
  return {
    ...(lhs?.hideIfNoChildren && rhs?.hideIfNoChildren ? { hideIfNoChildren: true } : undefined),
    ...(lhs?.hideInHierarchy && rhs?.hideInHierarchy ? { hideInHierarchy: true } : undefined),
    ...(lhs?.groupByClass || rhs?.groupByClass ? { groupByClass: true } : undefined),
    ...(lhs?.groupByLabel || rhs?.groupByLabel ? { groupByLabel: true } : undefined),
    ...(lhs?.mergeByLabelId ? { mergeByLabelId: lhs.mergeByLabelId } : undefined),
  };
}

function mergeNodeKeys(lhs: HierarchyNodeKey, rhs: HierarchyNodeKey): HierarchyNodeKey {
  if (HierarchyNodeKey.isCustom(lhs) && HierarchyNodeKey.isCustom(rhs)) {
    assert(lhs === rhs);
    return lhs;
  }
  assert(HierarchyNodeKey.isStandard(lhs) && HierarchyNodeKey.isStandard(rhs) && lhs.type === rhs.type);
  if (HierarchyNodeKey.isInstances(lhs)) {
    assert(HierarchyNodeKey.isInstances(rhs));
    return { type: "instances", instanceKeys: [...lhs.instanceKeys, ...rhs.instanceKeys] };
  }
  if (HierarchyNodeKey.isClassGrouping(lhs)) {
    assert(HierarchyNodeKey.isClassGrouping(rhs));
    assert(lhs.class.id === rhs.class.id);
    return { ...lhs };
  }
  if (HierarchyNodeKey.isLabelGrouping(lhs)) {
    assert(HierarchyNodeKey.isLabelGrouping(rhs));
    assert(lhs.label === rhs.label);
    return { ...lhs };
  }
  throw new Error(`Unable to merge given node keys`);
}

/** @internal */
export function mergeNodes(lhs: HierarchyNode, rhs: HierarchyNode): HierarchyNode {
  const mergedParams = mergeNodeHandlingParams(lhs.params, rhs.params);
  return {
    label: lhs.label,
    key: mergeNodeKeys(lhs.key, rhs.key),
    children:
      Array.isArray(lhs.children) && Array.isArray(rhs.children)
        ? [...lhs.children, ...rhs.children]
        : lhs.children === true || rhs.children === true
        ? true
        : lhs.children === false && rhs.children === false
        ? false
        : undefined,
    ...(lhs.autoExpand || rhs.autoExpand ? { autoExpand: lhs.autoExpand || rhs.autoExpand } : undefined),
    ...(lhs.extendedData || rhs.extendedData ? { extendedData: { ...lhs.extendedData, ...rhs.extendedData } } : undefined),
    ...(mergedParams ? { params: mergedParams } : undefined),
  };
}

/** @internal */
export function hasChildren<TNode extends { children?: boolean | Array<unknown> }>(node: TNode) {
  return node.children === true || (Array.isArray(node.children) && node.children.length > 0);
}

/** @internal */
export function mergeNodesObs(lhs: HierarchyNode, rhs: HierarchyNode, directNodesCache: Map<string, Observable<HierarchyNode>>) {
  const merged = mergeNodes(lhs, rhs);
  mergeDirectNodeObservables(lhs, rhs, merged, directNodesCache);
  return merged;
}

/** @internal */
export function mergeDirectNodeObservables(a: HierarchyNode, b: HierarchyNode, m: HierarchyNode, cache: Map<string, Observable<HierarchyNode>>) {
  const cachedA = cache.get(JSON.stringify(a.key));
  if (!cachedA) {
    return;
  }
  const cachedB = cache.get(JSON.stringify(b.key));
  if (!cachedB) {
    return;
  }
  const merged = merge(cachedA, cachedB);
  cache.set(JSON.stringify(m.key), merged);
}

/** @internal */
export function createOperatorLoggingNamespace(operatorName: string) {
  return `Presentation.HierarchyBuilder.Operators.${operatorName}`;
}
