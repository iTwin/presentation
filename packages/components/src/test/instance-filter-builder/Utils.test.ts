/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/

import { expect } from "chai";
import sinon from "sinon";
import { PropertyDescription, PropertyValueFormat, StandardTypeNames } from "@itwin/appui-abstract";
import { PropertyFilterRuleOperator } from "@itwin/components-react";
import { EmptyLocalization } from "@itwin/core-common";
import { IModelConnection } from "@itwin/core-frontend";
import { Descriptor, NavigationPropertyInfo } from "@itwin/presentation-common";
import { Presentation } from "@itwin/presentation-frontend";
import { renderHook } from "@testing-library/react";
import {
  createInstanceFilterPropertyInfos,
  DEFAULT_ROOT_CATEGORY_NAME,
  filterRuleValidator,
  INSTANCE_FILTER_FIELD_SEPARATOR,
  useFilterBuilderNavigationPropertyEditorContext,
} from "../../presentation-components/instance-filter-builder/Utils";
import { createTestECClassInfo } from "../_helpers/Common";
import {
  createTestCategoryDescription,
  createTestContentDescriptor,
  createTestNestedContentField,
  createTestPropertiesContentField,
  createTestSimpleContentField,
} from "../_helpers/Content";

describe("createInstanceFilterPropertyInfos", () => {
  it("creates property infos when fields are in root category", () => {
    const rootCategory = createTestCategoryDescription({ name: "root", label: "Root Category" });
    const descriptor = createTestContentDescriptor({
      categories: [rootCategory],
      fields: [
        createTestPropertiesContentField({
          properties: [{ property: { classInfo: createTestECClassInfo(), name: "prop1", type: "string" } }],
          category: rootCategory,
        }),
        createTestPropertiesContentField({
          properties: [{ property: { classInfo: createTestECClassInfo(), name: "prop2", type: "number" } }],
          category: rootCategory,
        }),
      ],
    });

    const input = createInstanceFilterPropertyInfos(descriptor);
    expect(input).to.matchSnapshot();
  });

  it("creates property info with default root category name and does not assign a label to it", () => {
    const rootCategory = createTestCategoryDescription({ name: DEFAULT_ROOT_CATEGORY_NAME, label: "Root Category" });
    const descriptor = createTestContentDescriptor({
      categories: [rootCategory],
      fields: [
        createTestPropertiesContentField({
          properties: [{ property: { classInfo: createTestECClassInfo(), name: "prop1", type: "string" } }],
          category: rootCategory,
        }),
      ],
    });

    const propertyInfos = createInstanceFilterPropertyInfos(descriptor);
    expect(propertyInfos[0].categoryLabel).to.be.undefined;
  });

  it("creates property infos when fields are in different categories category", () => {
    const rootCategory = createTestCategoryDescription({ name: "root", label: "Root Category" });
    const nestedCategory1 = createTestCategoryDescription({ name: "nested1", label: "Nested Category 1", parent: rootCategory });
    const nestedCategory2 = createTestCategoryDescription({ name: "nested2", label: "Nested Category 2", parent: rootCategory });
    const nestedCategory21 = createTestCategoryDescription({ name: "nested21", label: "Nested Category 2 1", parent: nestedCategory2 });
    const descriptor = createTestContentDescriptor({
      categories: [rootCategory, nestedCategory1, nestedCategory2, nestedCategory21],
      fields: [
        createTestPropertiesContentField({
          properties: [{ property: { classInfo: createTestECClassInfo(), name: "prop1", type: "string" } }],
          category: nestedCategory1,
        }),
        createTestPropertiesContentField({
          properties: [{ property: { classInfo: createTestECClassInfo(), name: "prop2", type: "number" } }],
          category: nestedCategory21,
        }),
      ],
    });

    const input = createInstanceFilterPropertyInfos(descriptor);
    expect(input).to.matchSnapshot();
  });

  it("creates property infos when property fields are in nested fields", () => {
    const rootCategory = createTestCategoryDescription({ name: "root", label: "Root Category" });
    const propertyField1 = createTestPropertiesContentField({
      properties: [{ property: { classInfo: createTestECClassInfo(), name: "prop1", type: "string" } }],
      category: rootCategory,
    });
    const propertyField2 = createTestPropertiesContentField({
      properties: [{ property: { classInfo: createTestECClassInfo(), name: "prop2", type: "number" } }],
      category: rootCategory,
    });

    const descriptor = createTestContentDescriptor({
      categories: [rootCategory],
      fields: [
        createTestNestedContentField({ nestedFields: [propertyField1], category: rootCategory }),
        createTestNestedContentField({ nestedFields: [propertyField2], category: rootCategory }),
      ],
    });

    const input = createInstanceFilterPropertyInfos(descriptor);
    expect(input).to.matchSnapshot();
  });

  it("creates property info with nested field content class name", () => {
    const rootCategory = createTestCategoryDescription({ name: "root", label: "Root Category" });
    const propertyField = createTestPropertiesContentField({
      properties: [{ property: { classInfo: createTestECClassInfo({ name: "Schema:PropClass " }), name: "prop1", type: "string" } }],
      category: rootCategory,
    });

    const descriptor = createTestContentDescriptor({
      categories: [rootCategory],
      fields: [
        createTestNestedContentField({
          nestedFields: [propertyField],
          category: rootCategory,
          contentClassInfo: createTestECClassInfo({ name: "Schema:RelatedClass" }),
        }),
      ],
    });

    const input = createInstanceFilterPropertyInfos(descriptor);
    expect(input[0].className).to.be.eq("Schema:RelatedClass");
  });
});

describe("filterRuleValidator", () => {
  const numericProperty: PropertyDescription = {
    displayLabel: "Numeric Prop",
    name: "numeric-prop",
    typename: StandardTypeNames.Double,
  };
  const quantityProperty: PropertyDescription = {
    displayLabel: "Quantity Prop",
    name: "quantity-prop",
    typename: StandardTypeNames.Double,
    quantityType: "TestKOQ",
  };

  before(() => {
    sinon.stub(Presentation, "localization").get(() => new EmptyLocalization());
  });

  after(() => {
    sinon.restore();
  });

  it("returns error message for invalid numeric rule", () => {
    expect(
      filterRuleValidator({
        id: "test-id",
        groupId: "test-group-id",
        property: numericProperty,
        operator: PropertyFilterRuleOperator.IsEqual,
        value: {
          valueFormat: PropertyValueFormat.Primitive,
          value: undefined,
          displayValue: "Invalid",
        },
      }),
    ).to.be.eq("instance-filter-builder.error-messages.not-a-number");
  });

  it("returns error message for invalid quantity rule", () => {
    expect(
      filterRuleValidator({
        id: "test-id",
        groupId: "test-group-id",
        property: quantityProperty,
        operator: PropertyFilterRuleOperator.IsEqual,
        value: {
          valueFormat: PropertyValueFormat.Primitive,
          value: undefined,
          displayValue: "Invalid",
        },
      }),
    ).to.be.eq("instance-filter-builder.error-messages.invalid");
  });

  it("does not return error message for valid numeric rule", () => {
    expect(
      filterRuleValidator({
        id: "test-id",
        groupId: "test-group-id",
        property: numericProperty,
        operator: PropertyFilterRuleOperator.IsEqual,
        value: {
          valueFormat: PropertyValueFormat.Primitive,
          value: 10,
          displayValue: "10",
        },
      }),
    ).to.be.undefined;
  });

  it("does not return error message for valid quantity rule", () => {
    expect(
      filterRuleValidator({
        id: "test-id",
        groupId: "test-group-id",
        property: quantityProperty,
        operator: PropertyFilterRuleOperator.IsEqual,
        value: {
          valueFormat: PropertyValueFormat.Primitive,
          value: 10,
          displayValue: "10 unit",
        },
      }),
    ).to.be.undefined;
  });
});

describe("useFilterBuilderNavigationPropertyEditorContext", () => {
  interface Props {
    imodel: IModelConnection;
    descriptor: Descriptor;
  }
  const testImodel = {} as IModelConnection;

  it("returns navigation property info", async () => {
    const navigationPropertyInfo: NavigationPropertyInfo = {
      classInfo: { id: "2", label: "Prop Class", name: "TestSchema:PropClass" },
      targetClassInfo: { id: "3", label: "Target Class", name: "TestSchema:TargetClass" },
      isForwardRelationship: true,
      isTargetPolymorphic: true,
    };
    const fieldName = "field_name";
    const testDescriptor = createTestContentDescriptor({
      fields: [
        createTestPropertiesContentField({
          name: fieldName,
          properties: [
            {
              property: {
                classInfo: { id: "1", label: "Field Class", name: "TestSchema:FieldClass" },
                name: "nav_prop",
                type: "navigation",
                navigationPropertyInfo,
              },
            },
          ],
        }),
      ],
    });
    const propertyDescription: PropertyDescription = {
      displayLabel: "TestProp",
      name: `test_category${INSTANCE_FILTER_FIELD_SEPARATOR}${fieldName}`,
      typename: "navigation",
    };

    const { result } = renderHook(({ imodel, descriptor }: Props) => useFilterBuilderNavigationPropertyEditorContext(imodel, descriptor), {
      initialProps: { imodel: testImodel, descriptor: testDescriptor },
    });

    const info = await result.current.getNavigationPropertyInfo(propertyDescription);
    expect(info).to.be.deep.eq(navigationPropertyInfo);
  });

  it("returns `undefined` for non properties field", async () => {
    const fieldName = "field_name";
    const testDescriptor = createTestContentDescriptor({
      fields: [createTestSimpleContentField({ name: fieldName })],
    });
    const propertyDescription: PropertyDescription = {
      displayLabel: "TestProp",
      name: `test_category${INSTANCE_FILTER_FIELD_SEPARATOR}${fieldName}`,
      typename: "navigation",
    };

    const { result } = renderHook(({ imodel, descriptor }: Props) => useFilterBuilderNavigationPropertyEditorContext(imodel, descriptor), {
      initialProps: { imodel: testImodel, descriptor: testDescriptor },
    });

    const info = await result.current.getNavigationPropertyInfo(propertyDescription);
    expect(info).to.be.undefined;
  });
});
