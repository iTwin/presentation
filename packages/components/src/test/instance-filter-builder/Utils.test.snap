// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createInstanceFilterPropertyInfos creates property infos when fields are in different categories category 1`] = `
Array [
  Object {
    "categoryLabel": "Root Category | Nested Category 1",
    "className": "SchemaName:ClassName",
    "field": Object {
      "category": "nested1",
      "editor": undefined,
      "extendedData": undefined,
      "isReadonly": false,
      "label": "Properties Field",
      "name": "PropertiesField",
      "priority": 0,
      "properties": Array [
        Object {
          "property": Object {
            "classInfo": Object {
              "id": "0x1",
              "label": "Class Label",
              "name": "SchemaName:ClassName",
            },
            "name": "prop1",
            "type": "string",
          },
        },
      ],
      "renderer": undefined,
      "type": Object {
        "typeName": "string",
        "valueFormat": "Primitive",
      },
    },
    "propertyDescription": Object {
      "displayLabel": "Properties Field",
      "name": "root/nested1#PropertiesField",
      "typename": "string",
    },
    "sourceClassId": "0x1",
  },
  Object {
    "categoryLabel": "Root Category | Nested Category 2 | Nested Category 2 1",
    "className": "SchemaName:ClassName",
    "field": Object {
      "category": "nested21",
      "editor": undefined,
      "extendedData": undefined,
      "isReadonly": false,
      "label": "Properties Field",
      "name": "PropertiesField",
      "priority": 0,
      "properties": Array [
        Object {
          "property": Object {
            "classInfo": Object {
              "id": "0x1",
              "label": "Class Label",
              "name": "SchemaName:ClassName",
            },
            "name": "prop2",
            "type": "number",
          },
        },
      ],
      "renderer": undefined,
      "type": Object {
        "typeName": "string",
        "valueFormat": "Primitive",
      },
    },
    "propertyDescription": Object {
      "displayLabel": "Properties Field",
      "name": "root/nested2/nested21#PropertiesField",
      "typename": "string",
    },
    "sourceClassId": "0x1",
  },
]
`;

exports[`createInstanceFilterPropertyInfos creates property infos when fields are in root category 1`] = `
Array [
  Object {
    "categoryLabel": "Root Category",
    "className": "SchemaName:ClassName",
    "field": Object {
      "category": "root",
      "editor": undefined,
      "extendedData": undefined,
      "isReadonly": false,
      "label": "Properties Field",
      "name": "PropertiesField",
      "priority": 0,
      "properties": Array [
        Object {
          "property": Object {
            "classInfo": Object {
              "id": "0x1",
              "label": "Class Label",
              "name": "SchemaName:ClassName",
            },
            "name": "prop1",
            "type": "string",
          },
        },
      ],
      "renderer": undefined,
      "type": Object {
        "typeName": "string",
        "valueFormat": "Primitive",
      },
    },
    "propertyDescription": Object {
      "displayLabel": "Properties Field",
      "name": "root#PropertiesField",
      "typename": "string",
    },
    "sourceClassId": "0x1",
  },
  Object {
    "categoryLabel": "Root Category",
    "className": "SchemaName:ClassName",
    "field": Object {
      "category": "root",
      "editor": undefined,
      "extendedData": undefined,
      "isReadonly": false,
      "label": "Properties Field",
      "name": "PropertiesField",
      "priority": 0,
      "properties": Array [
        Object {
          "property": Object {
            "classInfo": Object {
              "id": "0x1",
              "label": "Class Label",
              "name": "SchemaName:ClassName",
            },
            "name": "prop2",
            "type": "number",
          },
        },
      ],
      "renderer": undefined,
      "type": Object {
        "typeName": "string",
        "valueFormat": "Primitive",
      },
    },
    "propertyDescription": Object {
      "displayLabel": "Properties Field",
      "name": "root#PropertiesField",
      "typename": "string",
    },
    "sourceClassId": "0x1",
  },
]
`;

exports[`createInstanceFilterPropertyInfos creates property infos when property fields are in nested fields 1`] = `
Array [
  Object {
    "categoryLabel": "Root Category",
    "className": "SchemaName:ClassName",
    "field": Object {
      "category": "root",
      "editor": undefined,
      "extendedData": undefined,
      "isReadonly": false,
      "label": "Properties Field",
      "name": "PropertiesField",
      "priority": 0,
      "properties": Array [
        Object {
          "property": Object {
            "classInfo": Object {
              "id": "0x1",
              "label": "Class Label",
              "name": "SchemaName:ClassName",
            },
            "name": "prop1",
            "type": "string",
          },
        },
      ],
      "renderer": undefined,
      "type": Object {
        "typeName": "string",
        "valueFormat": "Primitive",
      },
    },
    "propertyDescription": Object {
      "displayLabel": "Properties Field",
      "name": "root#NestedContentField$PropertiesField",
      "typename": "string",
    },
    "sourceClassId": "0x2",
  },
  Object {
    "categoryLabel": "Root Category",
    "className": "SchemaName:ClassName",
    "field": Object {
      "category": "root",
      "editor": undefined,
      "extendedData": undefined,
      "isReadonly": false,
      "label": "Properties Field",
      "name": "PropertiesField",
      "priority": 0,
      "properties": Array [
        Object {
          "property": Object {
            "classInfo": Object {
              "id": "0x1",
              "label": "Class Label",
              "name": "SchemaName:ClassName",
            },
            "name": "prop2",
            "type": "number",
          },
        },
      ],
      "renderer": undefined,
      "type": Object {
        "typeName": "string",
        "valueFormat": "Primitive",
      },
    },
    "propertyDescription": Object {
      "displayLabel": "Properties Field",
      "name": "root#NestedContentField$PropertiesField",
      "typename": "string",
    },
    "sourceClassId": "0x2",
  },
]
`;
